File: python/knn.py
"""
k nearest neighbour algorithm
"""

from numpy import array 

def knn_classify(target, record_set, labels, k):
    rows,columns = record_set.shape # get record_set's rows and columns
    distances = {}
    for i in xrange(rows):
        # calculate distance of target with every record
        # use Euclidian Distance
        sum = 0
        for j in xrange(columns):
            sum += pow(target[j]-record_set[i][j],2)
        score = pow(sum, 0.5)
        distances[i] = score
    # get the k few labels
    target_labels = {}
    count = 0
    if k>rows:
        k = rows
    for item in sorted(distances.iteritems(),          \
                        key=lambda(key,value):(value,key),  \
                        reverse=True):
        if count == k:
            break
        label = labels[item[0]]
        if target_labels.has_key(label):
            target_labels[label] += 1
        else:
            target_labels[label] = 0
        count += 1
    max_value = max(target_labels.values())
    return filter(lambda item: item[1] == max_value, target_labels.items())


def make_record_set():
    group = array([
        [1.0, 1.1],
        [1.0, 1.0],
        [0, 0],
        [0, 0.1]
    ])
    labels = ['A', 'A', 'B', 'B']
    return group, labels

group, labels = make_record_set()
target = [0,0]
print knn_classify(target, group, labels, 3)

File: perl/knn.pl
use warnings;
use strict 'vars';

use Smart::Comments;
use List::Util qw/max sum/;

sub knn_classfity {
    my ($target, $recordset, $labels, $k) = @_;

    my $rows      = scalar @{$recordset};
    my $columns   = scalar @{$recordset->[0]};
    my $distances = {};

    foreach my $i (0 .. $rows-1) {
        my $value = sum(map {
            ($target->[$_]-$recordset->[$i][$_]) ** 2
        } (0 .. $columns-1));
        $distances->{$i} = $value ** 0.5;
    }
    ### DISTANCES: %$distances
    $k = $rows if $k>$rows;
    ### $k
    my @k_indexs = (reverse sort {
                          $distances->{$a} <=> $distances->{$b}
                    } keys %$distances) [0 .. $k-1];
    ### @k_indexs
    my %k_labels    = ();
    foreach my $index (@k_indexs) {
        $k_labels{$labels->[$index]} += 1;
    }
    ### %k_labels
    my $max_labels  = max(values %k_labels);
    print $max_labels,"\n";
    return grep {$k_labels{$_} == $max_labels} keys %k_labels;
}

sub make_record_set {
    my $record_set = [
        [1.0, 1.1],
        [1.0, 1.0],
        [0, 0],
        [0, 0.1],
    ];
    my $labels = ['A', 'A', 'B', 'B'];
    return ($record_set, $labels);
}


my ($group, $labels) = make_record_set();
my $target = [0,0];
print knn_classfity($target, $group, $labels, 3), "\n";
